tools/swarm/src/schema.rs

```diff
@@ -204,17 +204,66 @@ impl<'a> GenesisEnv<'a> {
 #[derive(Debug)]
 struct PortMapping(u16, u16);
 
+#[derive(Copy, Clone, Debug)]
+struct Filename<'a>(&'a str);
+
+impl std::fmt::Display for Filename<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0)
+    }
+}
+
+/// Path on the host.
+#[derive(Copy, Clone, Debug)]
+struct HostFile<'a>(&'a path::RelativePath, Filename<'a>);
+
+impl std::fmt::Display for HostFile<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{}{}", self.0.as_ref().display(), self.1))
+    }
+}
+
 /// Path inside the container.
-#[derive(serde::Serialize, Copy, Clone, Debug)]
-#[serde(transparent)]
+#[derive(Copy, Clone, Debug)]
 struct ContainerPath<'a>(&'a str);
 
-const CONTAINER_CONFIG_DIR: &str = "/config";
-const GENESIS_FILE: &str = "/tmp/genesis.signed.scale";
+impl std::fmt::Display for ContainerPath<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0)
+    }
+}
+
+/// Path inside the container.
+#[derive(Copy, Clone, Debug)]
+struct ContainerFile<'a>(ContainerPath<'a>, Filename<'a>);
+
+impl std::fmt::Display for ContainerFile<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{}{}", self.0, self.1))
+    }
+}
+
+const GENESIS_FILE: Filename = Filename("genesis.json");
+const CONFIG_FILE: Filename = Filename("client.toml");
+const GENESIS_SIGNED_SCALE: Filename = Filename("genesis.signed.scale");
+
+const CONTAINER_CONFIG_DIR: ContainerPath = ContainerPath("/config/");
+const CONTAINER_TMP_DIR: ContainerPath = ContainerPath("/tmp/");
+
+const CONTAINER_GENESIS_CONFIG: ContainerFile = ContainerFile(CONTAINER_CONFIG_DIR, GENESIS_FILE);
+const CONTAINER_CLIENT_CONFIG: ContainerFile = ContainerFile(CONTAINER_CONFIG_DIR, CONFIG_FILE);
+const CONTAINER_SIGNED_GENESIS: ContainerFile =
+    ContainerFile(CONTAINER_TMP_DIR, GENESIS_SIGNED_SCALE);
+
+#[derive(Copy, Clone, Debug)]
+struct ReadOnly;
 
 /// Mapping between `host:container` paths.
 #[derive(Copy, Clone, Debug)]
-struct PathMapping<'a>(HostPath<'a>, ContainerPath<'a>);
+struct PathMapping<'a>(HostFile<'a>, ContainerFile<'a>, ReadOnly);
```

Comment:
Why do you need `ReadOnly` struct? should `bool` be used instead to indicate whatever file is read only or not?

Reply:
We don't need it, but it makes it explicit that `PathMapping` is always read-only for now.